blockchain/v2/scheduler.go

```diff
@@ -0,0 +1,548 @@
+}
+
+func NewScheduler(minHeight int64, minRecvRate int64, peerTimeout time.Duration,
+	targetPending uint32, targetReceived uint32) *Scheduler {
+	return &Scheduler{
+		sc:             newSchedule(minHeight),
+		minRecvRate:    minRecvRate,
+		peerTimeout:    peerTimeout,
+		targetPending:  targetPending,
+		targetReceived: targetReceived,
+	}
+}
+
+type Skip struct {
+}
+
+func (sdr *Scheduler) handleAddPeer(peerID p2p.ID) Event {
+	err := sdr.sc.addPeer(peerID)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	return Skip{}
+}
+
+type Events []Event
+
+func (sdr *Scheduler) handleRemovePeer(peerID p2p.ID) Event {
+	err := sdr.sc.removePeer(peerID)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	return Skip{}
+}
+
+func (sdr *Scheduler) handleStatusResponse(peerID p2p.ID, height int64, now time.Time) Event {
+	err := sdr.sc.touchPeer(peerID, now)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	err = sdr.sc.setPeerHeight(peerID, height)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	return Skip{}
+}
+
+type bcBlockResponseEv struct {
+	peerID  p2p.ID
+	height  int64
+	block   *types.Block
+	msgSize int64
+}
+
+type scBlockReceivedEv struct {
+	peerID p2p.ID
+}
+
+func (sdr *Scheduler) handleBlockResponse(peerID p2p.ID, msg *bcBlockResponseEv, now time.Time) Event {
+	err := sdr.sc.touchPeer(peerID, now)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	err = sdr.sc.markReceived(peerID, msg.height, msg.msgSize, now)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	return scBlockReceivedEv{peerID}
+}
+
+type scFinishedEv struct{}
+
+func (sdr *Scheduler) handleBlockProcessed(peerID p2p.ID, height int64) Event {
+	err := sdr.sc.markProcessed(height)
+	if err != nil {
+		return schedulerErrorEv{peerID, err}
+	}
+
+	if sdr.sc.allBlocksProcessed() {
+		return scFinishedEv{}
+	}
+
+	return Skip{}
+}
+
+type scPrunePeerEv struct {
+	peerID p2p.ID
+	reason error
+}
+
+func (sdr *Scheduler) handleBlockProcessError(peerID p2p.ID, height int64) Event {
+	err := sdr.sc.removePeer(peerID)
+	if err != nil {
+		return scPrunePeerEv{peerID: peerID, reason: fmt.Errorf("Failed to process block %d", height)}
+	}
+
+	return Skip{}
+}
+
+type scSchedulerFailure struct {
+	peerID p2p.ID
+	time   time.Time
+	reason error
+}
+
+func (sdr *Scheduler) handleTimeCheck(now time.Time) Events {
+	prunablePeers := sdr.sc.prunablePeers(sdr.peerTimeout, sdr.minRecvRate, now)
+
+	events := []Event{}
+	for _, peerID := range prunablePeers {
+		err := sdr.sc.removePeer(peerID)
+		if err != nil {
+			events = append(events, scPrunePeerEv{peerID: peerID, reason: err})
+		}
+	}
+
+	pendingBlocks := sdr.sc.numBlockInState(blockStatePending)
+	receivedBlocks := sdr.sc.numBlockInState(blockStateReceived)
+	todo := math.Min(float64(sdr.targetPending-pendingBlocks), float64(sdr.targetReceived-receivedBlocks))
+	for height := sdr.sc.minHeight(); height <= sdr.sc.maxHeight(); height++ {
```

Comment:
Maybe we can avoid computing maxHeight() in this loop.
```suggestion
      maxHeight := sdr.sc.maxHeight()
	for height := sdr.sc.minHeight(); height <= maxHeight; height++ {
```